components/places/src/storage/bookmarks.rs

```diff
@@ -1241,14 +1248,31 @@ pub fn fetch_tree(
                     children: Vec::new(),
                 }
                 .into(),
-                BookmarkType::Bookmark => BookmarkNode {
-                    guid: Some(row.guid.clone()),
-                    date_added: Some(row.date_added),
-                    last_modified: Some(row.last_modified),
-                    title: row.title,
-                    url: Url::parse(row.url.unwrap().as_str())?,
+                BookmarkType::Bookmark => {
+                    // pretend invalid or missing URLs don't exist.
+                    match row.url {
+                        Some(str_val) => match Url::parse(str_val.as_str()) {
+                            // an invalid URL presumably means a logic error
+                            // somewhere far away from here...
+                            Err(_) => return Ok(None),
+                            Ok(url) => BookmarkNode {
+                                guid: Some(row.guid.clone()),
+                                date_added: Some(row.date_added),
+                                last_modified: Some(row.last_modified),
+                                title: row.title,
+                                url,
+                            }
+                            .into(),
+                        },
+                        // This is double-extra-invalid because various
+                        // constaints in the schema should prevent it - so make
+                        // some noise as it probably means something extra-bad?
```

Comment:
FWIW, on-disk database corruption could be one of the cases. On Desktop, there’s a very low but persistent occurrence of things that ought to be impossible in the schema, like `guid`s being null, types being totally nonsensical (what’s a bookmark type of -47231?), duplicate GUIDs in the tree, `moz_places` or `moz_bookmarks` not existing at all...

Anyway, I totally agree about making noise in this case, and also meaning something extra-bad! 😆